File: backend/src/validators/schemas.ts
import { z } from 'zod';

export const registerSchema = z.object({
  name: z.string().min(2, 'Name must be at least 2 characters').max(80),
  email: z.string().email('A valid email is required'),
  password: z.string().min(6, 'Password must be at least 6 characters').max(128),
  businessType: z.string().optional(),
});

export const loginSchema = z.object({
  email: z.string().email('A valid email is required'),
  password: z.string().min(1, 'Password is required'),
});

export const searchSchema = z.object({
  category: z.string().min(1, 'Category is required'),
  suppliers: z.array(z.string()).optional().default([]),
  query: z.string().min(1, 'Search query is required'),
  sortBy: z.enum(['lowest_price', 'highest_rating', 'fastest_delivery', 'highest_discount']).optional(),
  filters: z
    .object({
      brand: z.string().optional(),
      supplier: z.string().optional(),
      minRating: z.number().min(0).max(5).optional(),
      maxPrice: z.number().min(0).optional(),
      inStockOnly: z.boolean().optional(),
    })
    .optional(),
  weightProfile: z.enum(['balanced', 'startup', 'hospital', 'restaurant']).optional(),
});

export const recommendationSchema = z.object({
  products: z.array(z.any()).min(1, 'At least one product is required'),
  weightProfile: z.enum(['balanced', 'startup', 'hospital', 'restaurant']).optional(),
});

export const preferenceSchema = z.object({
  defaultCategory: z.string().optional(),
  enabledSuppliers: z.array(z.string()).optional(),
  sortPreference: z.enum(['lowest_price', 'highest_rating', 'fastest_delivery', 'highest_discount']).optional(),
  weightProfile: z.enum(['balanced', 'startup', 'hospital', 'restaurant']).optional(),
  businessType: z.string().optional(),
});

export const basketSchema = z.object({
  category: z.string().min(1, 'Category is required'),
  suppliers: z.array(z.string()).optional().default([]),
  items: z
    .array(
      z.object({
        query: z.string().min(1, 'Item name is required'),
        quantity: z.number().int().positive().max(999).optional(),
      }),
    )
    .min(1, 'Add at least one item to your basket'),
  weightProfile: z.enum(['balanced', 'startup', 'hospital', 'restaurant']).optional(),
  consolidationPenalty: z.number().min(0).max(100000).optional(),
});

File: backend/app/schemas.py
"""
Pydantic models for request validation — mirrors the Zod schemas from the
Node.js backend so the API contract stays identical.
"""
from typing import Optional, Literal

from pydantic import BaseModel, EmailStr, Field


# ---------------------------------------------------------------------------
# Auth
# ---------------------------------------------------------------------------

class RegisterInput(BaseModel):
    name: str = Field(..., min_length=2, max_length=80)
    email: EmailStr
    password: str = Field(..., min_length=6, max_length=128)
    businessType: Optional[str] = None


class LoginInput(BaseModel):
    email: EmailStr
    password: str = Field(..., min_length=1)


# ---------------------------------------------------------------------------
# Search
# ---------------------------------------------------------------------------

class SearchFilters(BaseModel):
    brand: Optional[str] = None
    supplier: Optional[str] = None
    minRating: Optional[float] = Field(None, ge=0, le=5)
    maxPrice: Optional[float] = Field(None, ge=0)
    inStockOnly: Optional[bool] = None


RecommendationMode = Literal[
    "balanced", "lowest_cost", "lowest_risk",
    "fastest_delivery", "highest_reliability", "best_long_term_value",
]


class SearchInput(BaseModel):
    category: str = Field(..., min_length=1)
    suppliers: list[str] = Field(default_factory=list)
    query: str = Field(..., min_length=1)
    sortBy: Optional[Literal["lowest_price", "lowest_total_cost", "highest_rating", "fastest_delivery", "highest_discount"]] = None
    filters: Optional[SearchFilters] = None
    weightProfile: Optional[Literal["balanced", "budget", "urgent", "fast"]] = None
    recommendationMode: Optional[RecommendationMode] = None
    includeSupplierHub: Optional[bool] = True
    userCity: Optional[str] = None


class RecommendationInput(BaseModel):
    products: list[dict] = Field(..., min_length=1)
    weightProfile: Optional[Literal["balanced", "budget", "urgent", "fast"]] = None
    recommendationMode: Optional[RecommendationMode] = None


# ---------------------------------------------------------------------------
# Preferences
# ---------------------------------------------------------------------------

class PreferenceInput(BaseModel):
    defaultCategory: Optional[str] = None
    enabledSuppliers: Optional[list[str]] = None
    sortPreference: Optional[Literal["lowest_price", "lowest_total_cost", "highest_rating", "fastest_delivery", "highest_discount"]] = None
    weightProfile: Optional[Literal["balanced", "budget", "urgent", "fast"]] = None
    businessType: Optional[str] = None
    city: Optional[str] = None


# ---------------------------------------------------------------------------
# Basket
# ---------------------------------------------------------------------------

class BasketItem(BaseModel):
    query: str = Field(..., min_length=1)
    quantity: Optional[int] = Field(None, ge=1, le=999)


class BasketInput(BaseModel):
    category: str = Field(..., min_length=1)
    suppliers: list[str] = Field(default_factory=list)
    items: list[BasketItem] = Field(..., min_length=1)
    weightProfile: Optional[Literal["balanced", "budget", "urgent", "fast"]] = None
    consolidationPenalty: Optional[float] = Field(None, ge=0, le=100000)
    recommendationMode: Optional[RecommendationMode] = None
    includeSupplierHub: Optional[bool] = True
    userCity: Optional[str] = None


class CurrentBasketInput(BaseModel):
    category: str = Field(..., min_length=1)
    items: list[BasketItem] = Field(default_factory=list)
